src/modules/newspapers/controllers/newspapers.controller.ts

```diff
@@ -0,0 +1,158 @@
+import https from 'https';
+
+import {
+	Controller,
+	ForbiddenException,
+	Get,
+	Header,
+	Headers,
+	Param,
+	Query,
+	Req,
+	Res,
+} from '@nestjs/common';
+import { ApiTags } from '@nestjs/swagger';
+import archiver from 'archiver';
+import { mapLimit } from 'blend-promise-utils';
+import { Request, Response } from 'express';
+import { isNil } from 'lodash';
+
+import { IeObjectsController } from '~modules/ie-objects/controllers/ie-objects.controller';
+import { convertObjectToCsv } from '~modules/ie-objects/helpers/convert-objects-to-csv';
+import { convertObjectToXml } from '~modules/ie-objects/helpers/convert-objects-to-xml';
+import {
+	NEWSPAPER_MIME_TYPE_ALTO,
+	NEWSPAPER_MIME_TYPE_BROWSE_COPY,
+} from '~modules/newspapers/newspapers.consts';
+import { SessionUserEntity } from '~modules/users/classes/session-user';
+import { Permission } from '~modules/users/types';
+import { RequireAllPermissions } from '~shared/decorators/require-permissions.decorator';
+import { SessionUser } from '~shared/decorators/user.decorator';
+
+@ApiTags('Newspapers')
+@Controller('newspapers')
+export class NewspapersController {
+	constructor(private ieObjectsController: IeObjectsController) {}
+
+	@Get(':id/export/zip')
+	@Header('Content-Type', 'application/zip')
+	@RequireAllPermissions(Permission.DOWNLOAD_OBJECT)
+	public async downloadNewspaperAsZip(
+		@Param('id') id: string,
+		@Query('page') pageIndex: number | undefined,
+		@Headers('referer') referer: string,
+		@Req() request: Request,
+		@Res() res: Response,
+		@SessionUser() user: SessionUserEntity
+	): Promise<void> {
+		const limitedObjectMetadata = await this.ieObjectsController.getIeObjectById(
+			id,
+			referer,
+			request,
+			user
+		);
+
+		if (!limitedObjectMetadata) {
+			throw new ForbiddenException(
+				'Object not found or you do not have permission to see it'
+			);
+		}
+
+		if (limitedObjectMetadata.dctermsFormat !== 'newspaper') {
+			throw new ForbiddenException(
+				'This object does not appear to be a newspaper. Only public newspapers can be downloaded'
+			);
+		}
+
+		const zipEntries: { filename: string; type: 'url' | 'string'; value: string }[] = [];
+
+		// Extract images and alto xml urls from the object
+		const exportSinglePage = !isNil(pageIndex) && !isNaN(pageIndex);
+		const pagesToExport = exportSinglePage
+			? limitedObjectMetadata.pageRepresentations.slice(pageIndex, pageIndex + 1)
+			: limitedObjectMetadata.pageRepresentations;
+		pagesToExport.forEach((representations, pageRepresentationIndex) => {
```

Comment:
Can you add more comments or maybe split it up in some more chucks? The 3 nested forEach loops make it really hard to read.

Reply:
👍 fixed in commit on develop